marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ProgressListener.java

```diff
@@ -0,0 +1,202 @@
+		this.totalResults = totalResults;
+		for (Consumer<ProgressUpdate> consumer : consumers) {
+			this.consumers.add(consumer);
+		}
+	}
+
+	public ProgressListener withTotalResults(long totalResults) {
+		this.totalResults = totalResults;
+		return this;
+	}
+
+	public ProgressListener onProgressUpdate(Consumer<ProgressUpdate> consumer) {
+		this.consumers.add(consumer);
+		return this;
+	}
+
+	/**
+	 * Initializes the start time so that each ProgressUpdate knows how long it occurred after the job was started.
+	 *
+	 * @param queryBatcher
+	 */
+	@Override
+	public void initializeListener(QueryBatcher queryBatcher) {
+		startTime = System.currentTimeMillis();
+	}
+
+	/**
+	 * Batches arrive in random order, so a ProgressUpdate is created and sent to each Consumer only if the
+	 * value of "getJobResultsSoFar" on the QueryBatch exceeds the number of results seen so far.
+	 * <p>
+	 * For example, if there are 2 batches, and batch 2 is processed first by this listener followed by batch 1, a
+	 * ProgressUpdate is only created when batch 2 is processed.
+	 *
+	 * @param batch
+	 */
+	@Override
+	public void processEvent(QueryBatch batch) {
+		// resultsSoFar is an AtomicLong so it can be safely updated across many threads
+		final long jobResultsSoFar = batch.getJobResultsSoFar();
+		final long newResultsSoFar = this.resultsSoFar.updateAndGet(operand ->
+			jobResultsSoFar > operand ? jobResultsSoFar : operand
+		);
+		boolean resultsSoFarWasUpdated = jobResultsSoFar == newResultsSoFar;
+
+		if (resultsSoFarWasUpdated && consumers != null) {
+			double timeSoFar = ((double) System.currentTimeMillis() - startTime) / 1000;
+
+			/**
+			 * The initial totalResults may have been incorrectly set to a value lower than jobResultsSoFar; if this
+			 * occurs, use jobResultsSoFar as the value passed to the ProgressUpdate object. totalResults is not
+			 * updated though in case there's a need to know what the initial value was.
+			 */
+			long totalForThisUpdate = jobResultsSoFar > this.totalResults && this.totalResults > 0 ? jobResultsSoFar : this.totalResults;
+
+			ProgressUpdate progressUpdate = new SimpleProgressUpdate(batch, startTime, totalForThisUpdate, timeSoFar);
+
+			for (Consumer<ProgressUpdate> consumer : consumers) {
+				invokeConsumer(consumer, progressUpdate);
+			}
+		}
+	}
+
+	/**
+	 * Protected so that a subclass can override how a consumer is invoked, particularly how an exception is handled.
+	 *
+	 * @param consumer
+	 * @param progressUpdate
+	 */
+	protected void invokeConsumer(Consumer<ProgressUpdate> consumer, ProgressUpdate progressUpdate) {
+		try {
+			consumer.accept(progressUpdate);
+		} catch (Throwable t) {
+			logger.error("Exception thrown by a Consumer<ProgressUpdate> consumer: " + consumer + "; progressUpdate: " + progressUpdate, t);
+		}
+	}
+
+	/**
+	 * Captures data of interest for a progress update.
+	 */
+	public interface ProgressUpdate {
+
+		String getProgressAsString();
```

Comment:
@rjrudin A belated thought! Do we always expect a String here? Can we use generics here and return a generic object so that just like how SimpleProgressUpdate can return a String and other implementations of ProgressUpdates can return whatever they want! What do you think?

Apologies for having this comment so late.

Reply:
I'm holding off on this for now, only because I don't think there's a strong use case for ProgressUpdate as an interface itself. Using a different implementation of it requires subclassing ProgressListener and overriding a protected method. If a client does that, then can always write a Consumer<ProgressUpdate> that casts the ProgressUpdate to their own concrete implementation in order to access any other functionality.

Reply:
Makes sense! Thanks Rob!